pandas/io/formats/format.py

```diff
@@ -1099,41 +1099,33 @@ def save_to_buffer(
     """
     Perform serialization. Write to buf or return as string if buf is None.
     """
-    with get_buffer(buf, encoding=encoding) as f:
-        f.write(string)
-        if buf is None:
-            return f.getvalue()
-        return None
-
-
-@contextmanager
-def get_buffer(buf: Optional[FilePathOrBuffer[str]], encoding: Optional[str] = None):
-    """
-    Context manager to open, yield and close buffer for filenames or Path-like
-    objects, otherwise yield buf unchanged.
-    """
-    if buf is not None:
-        buf = stringify_path(buf)
-    else:
-        buf = StringIO()
-
-    if encoding is None:
-        encoding = "utf-8"
-    elif not isinstance(buf, str):
+    is_path = isinstance(buf, (str, os.PathLike))
+    if not (is_path or encoding is None):
         raise ValueError("buf is not a file name and encoding is specified.")
-
-    if hasattr(buf, "write"):
-        yield buf
-    elif isinstance(buf, str):
-        with open(buf, "w", encoding=encoding, newline="") as f:
-            # GH#30034 open instead of codecs.open prevents a file leak
-            #  if we have an invalid encoding argument.
-            # newline="" is needed to roundtrip correctly on
-            #  windows test_to_latex_filename
-            yield f
-    else:
+    if not (is_path or buf is None or hasattr(buf, "write")):
         raise TypeError("buf is not a file name and it has no write method")
 
+    should_close = buf is None
+    if should_close:
+        buf = StringIO()
+    assert buf is not None
+    handles = get_handle(buf, "w", encoding=encoding)
+    if should_close:
+        assert isinstance(buf, StringIO)
+        handles.created_handles.append(buf)
+    with handles:
+        # Argument 1 to "write" of "RawIOBase" has incompatible type "str";
+        # expected "Union[bytes, bytearray]"  [arg-type]
+        # Argument 1 to "write" of "BufferedIOBase" has incompatible type
+        # "str"; expected "Union[bytes, bytearray]" [arg-type]
+        # Argument 1 to "write" of "_mmap" has incompatible type "str";
+        # expected "bytes"  [arg-type]
+        handles.handle.write(string)  # type: ignore[arg-type]
```

Comment:
can you not assert that this is a string?

Reply:
I think it might be possible to partially bind the input buffer type of `get_handle` and its output using generics (string stays string (needed for this case), bytes might stay bytes, and bytes might be wrapped to string).

Reply:
I think the way to approach this typing issue would be to overload get_handle (one w/wo `is_text`) to then define a different return type when `is_text` is present (`IOHandlesStr`, otherwise `IOHanldesBytes`). I think the issue this PR is addressing a is quite a corner case (I couldn't find another example). I' happy to look into the typing part (which will take me some time), even if it doesn't make it for 1.2.

Reply:
I removed this part from the PR. Except for having two functions `get_handle_string` and `get_handle_bytes`, there is no good way to grantee the caller to get a string/bytes handles. Even if we have two get_handles, it adds more code while removing only very few mypy-ignore statements.